File: tests/xcalablemp/global-view/loop/F/gblock2.f90
program main
  include 'xmp_lib.h'
  integer,parameter:: N=1000
!$xmp nodes p(4,4)
!$xmp template t(N,N)
  integer procs, result
  integer a(N,N), sa, ans
  real*8  b(N,N), sb
  real*4  c(N,N), sc
  integer :: g(4) = (/50,50,50,850/)
!$xmp distribute t(gblock(g),block) onto p
!$xmp align a(i,j) with t(i,j)
!$xmp align b(i,j) with t(i,j)
!$xmp align c(i,j) with t(i,j)

  sa = 0
  sb = 0.0
  sc = 0.0
      
!$xmp loop (i,j) on t(i,j)
  do j=1, N
     do i=1, N
        a(i,j) = 1
        b(i,j) = 2.0
        c(i,j) = 3.0
     enddo
  enddo
  
!$xmp loop (i,j) on t(i,j)
  do j=1, N
     do i=1, N
        sa = sa + a(i,j)
        sb = sb + b(i,j)
        sc = sc + c(i,j)
     enddo
  enddo

  result = 0
  procs = xmp_node_num()
  ans = g( (mod((procs)-1,4)+1)) * (1000/4)

  if(sa.ne.ans) then
     result = -1
  endif
  if(abs(sb-2.0*dble(ans)).gt.0.000000001) then
     result = -1
  endif
  if(abs(sc-3*real(ans)).gt.0.000001) then
     result = -1
  endif

!$xmp reduction(+:result)
!$xmp task on p(1,1)
  if (result .eq. 0) then
     write(*,*) "PASS"
  else
     write(*,*) "ERROR"
     call exit(1)
  end if
!$xmp end task

end program main

         

File: tests/XMP/global-view/loop/C/gblock2.c
#include <xmp.h>
#include <stdio.h> 
#include <stdlib.h>     
#pragma xmp nodes p[4][4]
static const int N=1000;
int g[4] = {50,50,50,850};
#pragma xmp template t[N][N]
#pragma xmp distribute t[block][gblock(g)] onto p
int a[N][N],sa=0,ans;
double b[N][N],sb=0.0;
float c[N][N],sc=0.0;
#pragma xmp align a[i][j] with t[i][j]
#pragma xmp align b[i][j] with t[i][j]
#pragma xmp align c[i][j] with t[i][j]
int i,j,result=0;

int main(void)
{
#pragma xmp loop on t[i][j]
  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      a[i][j] = 1;
      b[i][j] = 2.0;
      c[i][j] = 3.0;
    }
  }

#pragma xmp loop on t[i][j]
  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      sa += a[i][j];
      sb += b[i][j];
      sc += c[i][j];
    }
  }

  ans = (1000/4) * g[(xmp_node_num()-1)%4];
  if(sa != ans){
    result = -1;
  }
  if(abs(sb-2.0*(double)ans)>0.000000001){
    result = -1;
  }
  if(abs(sc-3*(float)ans)>0.000001){
    result = -1;
  }

#pragma xmp reduction(+:result)
#pragma xmp task on p[0][0]
  {
    if(result == 0 ){
      printf("PASS\n");
    }
    else{
      fprintf(stderr, "ERROR\n");
      exit(1);
    }
  }

  return 0;
}    
